clippy_lints/src/types.rs

```diff
@@ -1449,3 +1451,314 @@ impl<'a, 'tcx> LateLintPass<'a, 'tcx> for InvalidUpcastComparisons {
+                        target.span(),
+                        &format!("impl for `{}` should be generarized over different hashers", target.type_name()),
+                        move |db| {
+                            suggestion(cx, db, generics.span, generics_suggestion_span, target, ctr_vis);
+                        },
+                    );
+                }
+            },
+            ItemFn(ref decl, .., ref generics, body_id) => {
+                if item.vis != Public {
+                    return;
+                }
+
+                let body = cx.tcx.hir.body(body_id);
+
+                for ty in &decl.inputs {
+                    let mut vis = ImplicitHasherTypeVisitor::new(cx);
+                    vis.visit_ty(ty);
+
+                    for target in &vis.found {
+                        let generics_suggestion_span = generics.span.substitute_dummy({
+                            let pos = snippet_opt(cx, item.span.until(body.arguments[0].pat.span))
+                                .and_then(|snip| {
+                                    let i = snip.find("fn")?;
+                                    Some(item.span.lo() + BytePos((i + (&snip[i..]).find('(')?) as u32))
+                                })
+                                .expect("failed to create span for type parameters");
+                            Span::new(pos, pos, item.span.data().ctxt)
+                        });
+
+                        let mut ctr_vis = ImplicitHasherConstructorVisitor::new(cx, target);
+                        ctr_vis.visit_body(body);
+                        assert!(ctr_vis.suggestions.is_empty());
+
+                        span_lint_and_then(
+                            cx,
+                            IMPLICIT_HASHER,
+                            target.span(),
+                            &format!(
+                                "parameter of type `{}` should be generarized over different hashers",
+                                target.type_name()
+                            ),
+                            move |db| {
+                                suggestion(cx, db, generics.span, generics_suggestion_span, target, ctr_vis);
+                            },
+                        );
+                    }
+                }
+            },
+            _ => {},
+        }
+    }
+}
+
+enum ImplicitHasherType<'tcx> {
+    HashMap(Span, Ty<'tcx>, Cow<'static, str>, Cow<'static, str>),
+    HashSet(Span, Ty<'tcx>, Cow<'static, str>),
+}
+
+impl<'tcx> ImplicitHasherType<'tcx> {
+    /// Checks that `ty` is a target type without a BuildHasher.
+    fn new<'a>(cx: &LateContext<'a, 'tcx>, hir_ty: &hir::Ty) -> Option<Self> {
+        if let TyPath(QPath::Resolved(None, ref path)) = hir_ty.node {
+            let params = &path.segments.last().as_ref()?.parameters.as_ref()?.types;
+            let params_len = params.len();
+
+            let ty = cx.tcx.type_of(opt_def_id(path.def)?);
```

Comment:
This returns the type of definition (like `HashMap<K, V, S>`), not what is used in the code.

Can I get the `ty::Ty` that corresponds to a `hir::Ty`? Or is there any way to test if a `ty::Ty` equals to a `hir::Ty`?

Reply:
Have you tried `cx.tables.node_id_to_ty(hir_ty.hir_id)`?

Reply:
No. I tried now but it panicked (`node_id_to_type: no type for node`).

Reply:
[`rustc_typeck::hir_ty_to_ty`](https://manishearth.github.io/rust-internals-docs/rustc_typeck/fn.hir_ty_to_ty.html) seems to work. However its doc page says "quasi-deprecated":(

Reply:
If it ever gets deprecated, we can move it to the `utils` module:

```rust
    // In case there are any projections etc, find the "environment"
    // def-id that will be used to determine the traits/predicates in
    // scope.  This is derived from the enclosing item-like thing.
    let env_node_id = tcx.hir.get_parent(hir_ty.id);
    let env_def_id = tcx.hir.local_def_id(env_node_id);
    let item_cx = self::collect::ItemCtxt::new(tcx, env_def_id);
    item_cx.to_ty(hir_ty)
```

Reply:
I see.